modules/ref_fb_module/src/fft_fb_impl.cpp

```diff
@@ -0,0 +1,234 @@
+                                      .build()));
+
+        const auto inputRange = inputDataDescriptor.getValueRange();
+        outputDataDescriptor = DataDescriptorBuilder()
+                                   .setSampleType(SampleType::Float64)
+                                   .setDimensions(dimensions)
+                                   .setUnit(inputDataDescriptor.getUnit())
+                                   .setValueRange(inputRange.assigned() ? Range(0, inputRange.getHighValue()) : Range(0, 10))
+                                   .build();
+
+        outputSignal.setDescriptor(outputDataDescriptor);
+
+        outputDomainDataDescriptor =
+            DataDescriptorBuilderCopy(inputDomainDataDescriptor).setRule(ExplicitDataRule()).setSampleType(SampleType::UInt64).build();
+        outputDomainSignal.setDescriptor(outputDomainDataDescriptor);
+        configValid = true;
+    }
+    catch (const std::exception& e)
+    {
+        LOG_W("FFT: Failed to set descriptor for signal: {}", e.what())
+        outputSignal.setDescriptor(nullptr);
+    }
+}
+
+void FFTFbImpl::processEventPacket(const EventPacketPtr& packet)
+{
+    if (packet.getEventId() == event_packet_id::DATA_DESCRIPTOR_CHANGED)
+    {
+        const DataDescriptorPtr dataDesc = packet.getParameters().get(event_packet_param::DATA_DESCRIPTOR);
+        const DataDescriptorPtr domainDesc = packet.getParameters().get(event_packet_param::DOMAIN_DATA_DESCRIPTOR);
+        processSignalDescriptorChanged(dataDesc, domainDesc);
+    }
+}
+
+void FFTFbImpl::calculate()
+{
+    std::scoped_lock lock(sync);
+    SizeT readAmount = linearReader.getAvailableCount();
+    const auto inputData = std::make_unique<float[]>(readAmount * blockSize);
+    const auto inputDomainData = std::make_unique<uint64_t[]>(readAmount * blockSize);
+
+    const auto status = linearReader.readWithDomain(inputData.get(), inputDomainData.get(), &readAmount);
+
+    if (configValid && readAmount > 0)
+    {
+        const auto outputDomainPacket = DataPacket(outputDomainDataDescriptor, readAmount);
+        const auto outputDomainData = static_cast<uint64_t*>(outputDomainPacket.getData());
+        memset(outputDomainData, 0, outputDomainPacket.getRawDataSize());
+
+        const auto outputPacket = DataPacketWithDomain(outputDomainPacket, outputDataDescriptor, readAmount);
+        const auto outputData = static_cast<double*>(outputPacket.getData());
+        memset(outputData, 0, outputPacket.getRawDataSize());
+
+        for (size_t blockIdx = 0; blockIdx < readAmount; blockIdx++)
+        {
+            kiss_fft_cfg cfg = kiss_fft_alloc(static_cast<int>(blockSize), 0, nullptr, nullptr);
+            const auto in = std::make_unique<kiss_fft_cpx[]>(blockSize);
+            const auto out = std::make_unique<kiss_fft_cpx[]>(blockSize);
+                        
+            for (size_t i = 0; i < blockSize; i++)
+            {
+                in[i].r = inputData[i];
```

Comment:
I am not sure where you move the inputData pointer, so you might always just be reading from the first block of data here?

Reply:
True, this is an error on my part. The block index is missing.